In [1]:
# importing the necessary libraries for data analysis
import pandas as pd
import numpy as np
import glob

In [2]:
# Combine all city data into one big table
city_files = glob.glob("Dataset/*.csv")

all_cities_data = []


for file_name in city_files:
    
    city_df = pd.read_csv(file_name)
    all_cities_data.append(city_df)
    print(f"Loaded: {file_name}")
    
    
combined_data = pd.concat(all_cities_data, ignore_index=True)


#Save the combined data to a new CSV file
#combined_data.to_csv("all_cities_combined.csv", index=False)


print(f"SUCCESS: Combined {len(city_files)} city files into one table with {len(combined_data)} total rows")

Loaded: Dataset\PRSA_Data_Changping_20130301-20170228.csv
Loaded: Dataset\PRSA_Data_Dongsi_20130301-20170228.csv
Loaded: Dataset\PRSA_Data_Guanyuan_20130301-20170228.csv
Loaded: Dataset\PRSA_Data_Shunyi_20130301-20170228.csv
SUCCESS: Combined 4 city files into one table with 140256 total rows


In [3]:
## preview of combined data

#df= pd.read_csv('all_cities_combined.csv')
#df

combined_data

,No,year,month,day,hour,PM2.5,PM10,SO2,NO2,CO,O3,TEMP,PRES,DEWP,RAIN,wd,WSPM,station
0,1,2013,3,1,0,3.0,6.0,13.0,7.0,300.0,85.0,-2.3,1020.8,-19.7,0.0,E,0.5,Changping
1,2,2013,3,1,1,3.0,3.0,6.0,6.0,300.0,85.0,-2.5,1021.3,-19.0,0.0,ENE,0.7,Changping
2,3,2013,3,1,2,3.0,3.0,22.0,13.0,400.0,74.0,-3.0,1021.3,-19.9,0.0,ENE,0.2,Changping
3,4,2013,3,1,3,3.0,6.0,12.0,8.0,300.0,81.0,-3.6,1021.8,-19.1,0.0,NNE,1.0,Changping
4,5,2013,3,1,4,3.0,3.0,14.0,8.0,300.0,81.0,-3.5,1022.3,-19.4,0.0,N,2.1,Changping
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
140251,35060,2017,2,28,19,27.0,72.0,8.0,92.0,800.0,16.0,10.3,1014.2,-12.4,0.0,W,1.8,Shunyi
140252,35061,2017,2,28,20,47.0,55.0,17.0,86.0,1100.0,19.0,9.8,1014.5,-9.9,0.0,NW,1.5,Shunyi
140253,35062,2017,2,28,21,18.0,28.0,4.0,30.0,500.0,64.0,9.1,1014.6,-12.7,0.0,NE,1.7,Shunyi
140254,35063,2017,2,28,22,18.0,20.0,9.0,33.0,500.0,59.0,7.1,1015.2,-13.2,0.0,WNW,1.8,Shunyi


**Basic Data Understanding:**

To get familiar with  dataset, I started by examining the columns and their corresponding data types. I also reviewed the descriptive statistics and checked for any missing values.

In [4]:
combined_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140256 entries, 0 to 140255
Data columns (total 18 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   No       140256 non-null  int64  
 1   year     140256 non-null  int64  
 2   month    140256 non-null  int64  
 3   day      140256 non-null  int64  
 4   hour     140256 non-null  int64  
 5   PM2.5    137203 non-null  float64
 6   PM10     138144 non-null  float64
 7   SO2      137195 non-null  float64
 8   NO2      135964 non-null  float64
 9   CO       131607 non-null  float64
 10  O3       136326 non-null  float64
 11  TEMP     140112 non-null  float64
 12  PRES     140115 non-null  float64
 13  DEWP     140109 non-null  float64
 14  RAIN     140114 non-null  float64
 15  wd       139474 non-null  object 
 16  WSPM     140141 non-null  float64
 17  station  140256 non-null  object 
dtypes: float64(11), int64(5), object(2)
memory usage: 19.3+ MB


In [5]:
combined_data.head()

,No,year,month,day,hour,PM2.5,PM10,SO2,NO2,CO,O3,TEMP,PRES,DEWP,RAIN,wd,WSPM,station
0,1,2013,3,1,0,3.0,6.0,13.0,7.0,300.0,85.0,-2.3,1020.8,-19.7,0.0,E,0.5,Changping
1,2,2013,3,1,1,3.0,3.0,6.0,6.0,300.0,85.0,-2.5,1021.3,-19.0,0.0,ENE,0.7,Changping
2,3,2013,3,1,2,3.0,3.0,22.0,13.0,400.0,74.0,-3.0,1021.3,-19.9,0.0,ENE,0.2,Changping
3,4,2013,3,1,3,3.0,6.0,12.0,8.0,300.0,81.0,-3.6,1021.8,-19.1,0.0,NNE,1.0,Changping
4,5,2013,3,1,4,3.0,3.0,14.0,8.0,300.0,81.0,-3.5,1022.3,-19.4,0.0,N,2.1,Changping


In [6]:
combined_data.columns

Index(['No', 'year', 'month', 'day', 'hour', 'PM2.5', 'PM10', 'SO2', 'NO2',
       'CO', 'O3', 'TEMP', 'PRES', 'DEWP', 'RAIN', 'wd', 'WSPM', 'station'],
      dtype='object')

**Description of dataset columns:**



Year – The calendar year when the air quality measurement was taken.

Month – The month of the measurement, useful for seasonal or monthly trend analysis.

Day – The day of the month when the reading was recorded.

Hour – The hour of the day (0–23), allowing analysis of hourly pollution patterns.

PM2.5 – Fine particulate matter smaller than 2.5 micrometers. These particles penetrate deep into the lungs and are linked to serious health risks.

PM10 – Particulate matter smaller than 10 micrometers, including dust, smoke, and coarse particles.

SO2 – Sulfur dioxide, a pollutant produced mainly from burning coal or oil; contributes to respiratory irritation and acid rain.

NO2 – Nitrogen dioxide, a toxic gas from combustion sources like vehicles and industry; contributes to smog and respiratory issues.

CO – Carbon monoxide, a colorless gas from incomplete combustion; harmful at high concentrations.

O3 – Ground-level ozone, formed by chemical reactions in sunlight; a major component of smog and harmful to lungs.

TEMP – Ambient air temperature (°C) at the time of measurement.

PRES – Atmospheric pressure (hPa), useful for understanding weather conditions influencing pollution.

DEWP – Dew point temperature (°C), indicating moisture content in the air.

RAIN – Rainfall amount, typically in millimeters; rainfall can reduce particulate pollution.

wd – Wind direction, indicating the compass direction from which the wind is blowing (e.g., NW, NNE).

WSPM – Wind speed (m/s), which affects pollutant dispersion.

station 


In [7]:
combined_data.shape

(140256, 18)

# **Inference :**


---


The dataset contains 14,0256 records with 18 columns, and while the station and Date columns are fully complete, several pollutant measurements show missing data. Key pollutants such as PM2.5, PM10, NH3, Toluene, and Xylene have a significant number of missing values, with Xylene having the least completeness.

In [8]:
City = combined_data['station'].value_counts()
print(f'Total number of cities in the dataset : {len(City)}')
City

Total number of cities in the dataset : 4


station
Changping    35064
Dongsi       35064
Guanyuan     35064
Shunyi       35064
Name: count, dtype: int64

In [9]:
combined_data['station'].unique()

array(['Changping', 'Dongsi', 'Guanyuan', 'Shunyi'], dtype=object)

In [10]:
print(f"There are  {combined_data.duplicated().sum()} duplicates in the dataset")

There are  0 duplicates in the dataset



### **Dataset Structure**

The dataset contains **140,256 rows and 18 columns**. It includes hourly air‑quality pollutant measurements, meteorological variables (TEMP, PRES, DEWP, RAIN, wind direction and speed), and station names that represent the monitoring locations. The date information is stored across **year, month, day, and hour** columns.

There are two main types of columns in the dataset:

-   **Categorical columns:** _station_, _wd_, and the combined _date_ fields when treated as a categorical timestamp (3 columns)
    
-   **Numerical columns:** All pollutant concentrations (PM2.5, PM10, SO2, NO2, CO, O3) and meteorological variables (TEMP, PRES, DEWP, RAIN, WSPM), along with the index-like _No_ column (13 columns)



### **Data Completeness**

The dataset is largely complete, with **no missing values** in the columns: **year, month, day, hour, station, No**

The pollutant and weather variables show varying levels of missing data:

-   **Least missing:** _DEWP (0.10%)_, _TEMP (0.10%)_, _RAIN (0.10%)_, _PRES (0.10%)_, _WSPM (0.08%)_, _wd (0.56%)_
    
-   **Moderate missing:** _PM10 (1.50%)_, _PM2.5 (2.18%)_, _SO2 (2.18%)_, _O3 (2.80%)_, _NO2 (3.06%)_
    
-   **Most missing:** **CO (6.17%)** — the variable with the highest proportion of missing values in the dataset


In [11]:
#checking Missing values
def missing_values_table(df):
    mis_val = df.isnull().sum()

  
    mis_val_percent = 100 * mis_val / len(df)

    mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)

    mis_val_table = mis_val_table.rename(columns={0: 'Missing Values', 1: '% of Total Values'})

    mis_val_table = mis_val_table.sort_values('% of Total Values', ascending=False)

    return mis_val_table

missing_values = missing_values_table(combined_data)
display(missing_values.style.background_gradient(cmap='Greens'))

,Missing Values,% of Total Values
CO,8649,6.166581
NO2,4292,3.060119
O3,3930,2.802019
SO2,3061,2.182438
PM2.5,3053,2.176734
PM10,2112,1.505818
wd,782,0.557552
DEWP,147,0.104808
TEMP,144,0.102669
RAIN,142,0.101243


In [12]:
missing_values_per_station = combined_data.groupby('station').apply(lambda x: x.isnull().sum())
print(missing_values_per_station)

           No  year  month  day  hour  PM2.5  PM10   SO2   NO2    CO    O3  \
station                                                                      
Changping   0     0      0    0     0    774   582   628   667  1521   604   
Dongsi      0     0      0    0     0    750   553   663  1601  3197   664   
Guanyuan    0     0      0    0     0    616   429   474   659  1753  1173   
Shunyi      0     0      0    0     0    913   548  1296  1365  2178  1489   

           TEMP  PRES  DEWP  RAIN   wd  WSPM  station  
station                                                
Changping    53    50    53    51  140    43        0  
Dongsi       20    20    20    20   78    14        0  
Guanyuan     20    20    20    20   81    14        0  
Shunyi       51    51    54    51  483    44        0  


C:\Users\HP\AppData\Local\Temp\ipykernel_16088\23882343.py:1: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  missing_values_per_station = combined_data.groupby('station').apply(lambda x: x.isnull().sum())



# **Inference**

The table indicates that missing data is **not evenly distributed across stations or pollutants**. While the temporal fields (_year, month, day, hour_) and the station identifiers are fully complete across all locations, the pollutant and meteorological variables show substantial variation in data completeness.

-   All stations have **complete spatial and temporal coverage**, meaning no gaps in the date‑time structure or station labels. This ensures that time‑series analysis can be performed without structural interruptions.
    
-   Missingness varies widely across pollutants.
    
    -   **PM2.5, PM10, SO2, NO2, CO, and O3** show large differences between stations. For example, **CO** has the highest missing counts across all stations (e.g., 3197 missing in Dongsi, 2178 in Shunyi), indicating inconsistent CO monitoring or frequent sensor downtime.
        
    -   **PM2.5 and PM10** also show substantial missingness, especially in Shunyi and Changping, suggesting irregular particulate monitoring or data transmission issues.
        
-   Meteorological variables (**TEMP, PRES, DEWP, RAIN, WSPM**) have relatively **low missingness**, but still show station‑specific patterns. For instance, Shunyi has notably higher missing values for _wd_ (483) compared to Dongsi or Guanyuan (around 78–81), indicating localized issues with wind‑direction sensors.
    
-   The **station-level variation** is clear:
    
    -   **Dongsi and Guanyuan** have fewer missing entries for most pollutants.
        
    -   **Shunyi** shows the highest missingness for several pollutants (SO2, NO2, CO, O3), pointing to more frequent measurement gaps.
        
    -   **Changping** has moderate missingness across pollutants but relatively stable meteorological data.
        
-   Overall, the dataset provides strong temporal consistency but **uneven pollutant monitoring across stations**. Pollutants such as **CO, PM2.5, PM10, SO2, and O3** require careful handling—either imputation, interpolation, or exclusion—before conducting reliable comparative or trend analyses across stations.

In [13]:
#missing values per station
print(combined_data.groupby('station').apply(lambda x: x.isnull().sum().sum()))

station
Changping    5166
Dongsi       7600
Guanyuan     5279
Shunyi       8523
dtype: int64


C:\Users\HP\AppData\Local\Temp\ipykernel_16088\334427061.py:2: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  print(combined_data.groupby('station').apply(lambda x: x.isnull().sum().sum()))



# **Inference**

Missing values vary noticeably across stations, showing uneven data quality.

-   **Shunyi**  has the highest number of missing entries (8,523), indicating major gaps in pollutant or weather measurements.
    
-   **Dongsi**  also shows substantial missingness (7,600), suggesting less reliable monitoring.
    
-   **Guanyuan**  and **Changping**  have fewer missing values (5,279 and 5,166), making their data comparatively more complete.
    
-   Overall, while all stations have full date‑time coverage, the **completeness of pollutant measurements differs sharply**, and stations with higher missingness will require more careful preprocessing before analysis.

In [14]:
combined_data['Date'] = pd.to_datetime(
    combined_data[['year', 'month', 'day']],
    dayfirst=True
)

#print(combined_data)

# Make a copy of original data BEFORE  cleaning
df_raw = combined_data.copy()

print(df_raw)

           No  year  month  day  hour  PM2.5  PM10   SO2   NO2      CO    O3  \
0           1  2013      3    1     0    3.0   6.0  13.0   7.0   300.0  85.0   
1           2  2013      3    1     1    3.0   3.0   6.0   6.0   300.0  85.0   
2           3  2013      3    1     2    3.0   3.0  22.0  13.0   400.0  74.0   
3           4  2013      3    1     3    3.0   6.0  12.0   8.0   300.0  81.0   
4           5  2013      3    1     4    3.0   3.0  14.0   8.0   300.0  81.0   
...       ...   ...    ...  ...   ...    ...   ...   ...   ...     ...   ...   
140251  35060  2017      2   28    19   27.0  72.0   8.0  92.0   800.0  16.0   
140252  35061  2017      2   28    20   47.0  55.0  17.0  86.0  1100.0  19.0   
140253  35062  2017      2   28    21   18.0  28.0   4.0  30.0   500.0  64.0   
140254  35063  2017      2   28    22   18.0  20.0   9.0  33.0   500.0  59.0   
140255  35064  2017      2   28    23   15.0  22.0  13.0  34.0   500.0  60.0   

        TEMP    PRES  DEWP  RAIN   wd  


# **Methods Used for Handling Missing Values**

To keep the dataset suitable for time‑series analysis, different imputation techniques were applied depending on how each variable behaves over time. Instead of relying on simple mean or median filling—which can distort temporal patterns—the methods were chosen to preserve natural trends and realistic pollutant behavior.

**Time‑based interpolation** was used for pollutants that change rapidly from hour to hour, while **monthly median values** were applied to variables with more stable long‑term patterns. For gases with related chemical behavior, imputation followed known **scientific relationships** to maintain consistency.

Overall, these methods help retain realistic pollutant dynamics and produce cleaner, more reliable data for further analysis.

In [15]:
combined_data['Date'] = pd.to_datetime(combined_data['Date'], dayfirst=True)
df = combined_data.sort_values(["station", "Date"])

cols_to_interp = ["PM2.5", "PM10", "CO","TEMP", "PRES", "DEWP", "RAIN", "WSPM"]

def interpolate_city(group):
    group = group.set_index("Date") #temporary index creation
    for col in cols_to_interp:
        group[col] = group[col].interpolate(
            method="time",
            limit=3,          # fill small gaps
            limit_direction="both"
        )
    return group.reset_index()

df = df.groupby("station", group_keys=False).apply(interpolate_city)
print(df[cols_to_interp].isna().sum())

PM2.5    1340
PM10      741
CO       5394
TEMP        0
PRES        0
DEWP        0
RAIN        0
WSPM        0
dtype: int64


C:\Users\HP\AppData\Local\Temp\ipykernel_16088\489416493.py:16: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  df = df.groupby("station", group_keys=False).apply(interpolate_city)


In [16]:
# Sort by Date only
df = df.sort_values("Date")

def interpolate_date(df):
    df2 = df.set_index("Date")
    for col in cols_to_interp:
        df2[col] = df2[col].interpolate(
            method="time",
            limit_direction="both"
        )
    return df2.reset_index()

df = interpolate_date(df)
print(df[cols_to_interp].isna().sum())

PM2.5    0
PM10     0
CO       0
TEMP     0
PRES     0
DEWP     0
RAIN     0
WSPM     0
dtype: int64


In [17]:
df.isnull().sum()


Date          0
No            0
year          0
month         0
day           0
hour          0
PM2.5         0
PM10          0
SO2        3061
NO2        4292
CO            0
O3         3930
TEMP          0
PRES          0
DEWP          0
RAIN          0
wd          782
WSPM          0
station       0
dtype: int64




The dataset now shows complete pollutant values several meteorological variables still contain gaps. that weather‑related measurements were less consistently recorded than pollutant data.

Overall, while pollutant fields are fully cleaned, the remaining missing values in meteorological variables will require additional imputation to ensure a fully reliable dataset for time‑series or modeling work.

In [28]:
# Extract month from Date column
df['Month'] = df['Date'].dt.month

# Columns that can use monthly median (numeric only)
cols_to_fill = ["TEMP", "PRES", "DEWP", "RAIN", "WSPM","SO2","NO2","O3"]

# Monthly median imputation for numeric columns
for col in cols_to_fill:
    df[col] = df.groupby('Month')[col].transform(lambda x: x.fillna(x.median()))

# Monthly mode imputation for wind direction (categorical)
df['wd'] = df.groupby('Month')['wd'].transform(
    lambda x: x.fillna(x.mode()[0] if not x.mode().empty else x.ffill())
)

# Check remaining missing values
print(df[["TEMP", "PRES", "DEWP", "RAIN", "WSPM","SO2","NO2","O3"]].isna().sum())

TEMP    0
PRES    0
DEWP    0
RAIN    0
WSPM    0
SO2     0
NO2     0
O3      0
dtype: int64


In [29]:
df.isnull().sum()


Date                 0
No                   0
year                 0
month                0
day                  0
hour                 0
PM2.5                0
PM10                 0
SO2                  0
NO2                  0
CO                   0
O3                   0
TEMP                 0
PRES                 0
DEWP                 0
RAIN                 0
wd                   0
WSPM                 0
station              0
Month                0
AQI_recalc           0
AQI_Bucket_recalc    0
dtype: int64

# **Inference: AQI Calculation using CPCB Standards**

---

**Air Quality Index (AQI)** has been calculated following the **Central Pollution Control Board (CPCB), India** guidelines. This method is widely used in India for public communication of air quality.

### Key Features of the Implementation:

- **Breakpoints**: Official CPCB PM2.5 and PM10 concentration breakpoints were used.
- **Rounding Rule**: As per CPCB guidelines, pollutant concentrations were **rounded to the nearest integer** before computing sub-indices.
- **Sub-Index Calculation**: Linear interpolation method applied within each breakpoint range.
- **Final AQI**: Taken as the **maximum** of the PM2.5 and PM10 sub-indices (dominant pollutant approach).
- **AQI Categories**: Good, Satisfactory, Moderate, Poor, Very Poor, and Severe.

This allows direct comparison of Beijing’s air quality with Indian AQI standards and provides a standardized way to interpret pollution levels.

**Reference:**
- CPCB National Air Quality Index (AQI) Report  
  https://cpcb.nic.in/displaypdf.php?id=bmF0aW9uYWwtYWlyLXF1YWxpdHktaW5kZXgvRklOQUwtUkVQT1JUX0FRSV8ucGRm

In [33]:


# PCB Breakpoints
pm25_breakpoints = [
    (0,   30,   0,   50),
    (31,  60,  51,  100),
    (61,  90, 101,  200),
    (91, 120, 201,  300),
    (121,250, 301,  400),
    (251,500, 401,  500),
]

pm10_breakpoints = [
    (0,    50,   0,   50),
    (51,  100,  51,  100),
    (101, 250, 101,  200),
    (251, 350, 201,  300),
    (351, 430, 301,  400),
    (431,1000, 401,  500),
]


# Robust Sub-index Function
def sub_index(conc, bps):
    if pd.isna(conc):
        return None
    
    conc = float(conc)
    
    if conc < 0:
        conc = 0
    
    max_conc = bps[-1][1]   
    if conc > max_conc:
        return bps[-1][3]   
    
    for low, high, i_low, i_high in bps:
        if low <= conc <= high:
            return (i_high - i_low) * (conc - low) / (high - low) + i_low
    
    return None  


# Round the concentrations (CPCB Rule)
df['PM2.5_round'] = df['PM2.5'].round().astype("Int64")
df['PM10_round']  = df['PM10'].round().astype("Int64")


# Compute Sub-indices
df['PM2.5_sub'] = df['PM2.5_round'].apply(lambda x: sub_index(x, pm25_breakpoints))
df['PM10_sub']  = df['PM10_round'].apply(lambda x: sub_index(x, pm10_breakpoints))


# Final AQI = Max of sub-indices
df['AQI_recalc'] = df[['PM2.5_sub', 'PM10_sub']].max(axis=1).round().astype('Int64')


# 6. AQI Bucket
def aqi_bucket_cpcb(aqi):
    if pd.isna(aqi):
        return None
    a = float(aqi)
    if a <= 50:
        return "Good"
    elif a <= 100:
        return "Satisfactory"
    elif a <= 200:
        return "Moderate"
    elif a <= 300:
        return "Poor"
    elif a <= 400:
        return "Very Poor"
    else:
        return "Severe"

df['AQI_Bucket_recalc'] = df['AQI_recalc'].apply(aqi_bucket_cpcb)



display(df[['Date', 'station', 'PM2.5', 'PM10', 
            'PM2.5_round', 'PM10_round', 
            'PM2.5_sub', 'PM10_sub', 
            'AQI_recalc', 'AQI_Bucket_recalc']].head(10))

,Date,station,PM2.5,PM10,PM2.5_round,PM10_round,PM2.5_sub,PM10_sub,AQI_recalc,AQI_Bucket_recalc
0,2013-03-01,Changping,3.0,6.0,3,6,5.000000,6.0,6,Good
1,2013-03-01,Dongsi,7.0,7.0,7,7,11.666667,7.0,12,Good
2,2013-03-01,Dongsi,4.0,4.0,4,4,6.666667,4.0,7,Good
3,2013-03-01,Dongsi,9.0,9.0,9,9,15.000000,9.0,15,Good
4,2013-03-01,Guanyuan,4.0,4.0,4,4,6.666667,4.0,7,Good
5,2013-03-01,Guanyuan,4.0,4.0,4,4,6.666667,4.0,7,Good
6,2013-03-01,Guanyuan,3.0,3.0,3,3,5.000000,3.0,5,Good
7,2013-03-01,Dongsi,3.0,3.0,3,3,5.000000,3.0,5,Good
8,2013-03-01,Guanyuan,3.0,6.0,3,6,5.000000,6.0,6,Good
9,2013-03-01,Guanyuan,3.0,6.0,3,6,5.000000,6.0,6,Good


In [34]:
#dropping temporary columns

df = df.drop(columns=[
    'PM2.5_round',
    'PM10_round',
    'PM2.5_sub',
    'PM10_sub'
])

In [32]:
df.isnull().sum()


Date                 0
No                   0
year                 0
month                0
day                  0
hour                 0
PM2.5                0
PM10                 0
SO2                  0
NO2                  0
CO                   0
O3                   0
TEMP                 0
PRES                 0
DEWP                 0
RAIN                 0
wd                   0
WSPM                 0
station              0
Month                0
AQI_recalc           0
AQI_Bucket_recalc    0
dtype: int64

In [ ]:
df.to_csv("AQI_cleaned_dataset.csv", index=False)
print(" Cleaned data saved to AQI_cleaned_dataset.csv")

 Cleaned data saved to AQI_cleaned_dataset.csv


In [42]:
df.describe().T

,count,mean,min,25%,50%,75%,max,std
Date,140256,2015-03-01 00:00:00,2013-03-01 00:00:00,2014-03-01 00:00:00,2015-03-01 00:00:00,2016-02-29 00:00:00,2017-02-28 00:00:00,NaN
No,140256.0,17532.5,1.0,8766.75,17532.5,26298.25,35064.0,10122.141
year,140256.0,2014.66256,2013.0,2014.0,2015.0,2016.0,2017.0,1.177201
month,140256.0,6.52293,1.0,4.0,7.0,10.0,12.0,3.448715
day,140256.0,15.729637,1.0,8.0,16.0,23.0,31.0,8.800123
hour,140256.0,11.5,0.0,5.75,11.5,17.25,23.0,6.922211
PM2.5,140256.0,79.969694,2.0,20.0,55.0,111.0,941.0,80.768871
PM10,140256.0,103.30964,2.0,35.0,81.0,142.0,999.0,91.401653
SO2,140256.0,15.984619,0.2856,2.0,7.0,20.0,310.0,21.740289
NO2,140256.0,49.727687,1.8477,25.0,43.0,68.0,270.0,32.603418


# **Inference:**

📊 **Summary Statistics — Interpretation**

📘 **Count**

All variables have **140,256** complete entries after data cleaning and imputation.  
This represents full hourly data for 4 Beijing stations over ~4 years (2013–2017), giving excellent temporal coverage.

---

📆 **Time Period**

- **Earliest date**: 2013-03-01  
- **Latest date**: 2017-02-28  
- **Median date**: 2015-03-01  

The dataset covers **4 full years** of hourly air quality measurements across Changping, Dongsi, Guanyuan, and Shunyi stations.

---

📈 **Mean Values (Average Pollution Levels)**

- **PM2.5** = **79.97 µg/m³** → Poor air quality on average (China standard for "Good" is ≤35).
- **PM10** = **103.31 µg/m³** → Frequently moderate to unhealthy levels.
- **SO₂** = **15.98 µg/m³** → Relatively low on average.
- **NO₂** = **49.73 µg/m³** → Moderate traffic-related pollution.
- **CO** = **1240 µg/m³** (1.24 mg/m³) → Generally acceptable but with spikes.
- **O₃** = **56.17 µg/m³** → Moderate ozone levels.

**Key Insight**: Particulate matter (**PM2.5 & PM10**) are the dominant pollutants in Beijing during this period, indicating serious air quality concerns.

---

📉 **Min & Max (Range Analysis)**

| Pollutant | Minimum     | Maximum    | Observation |
|-----------|-------------|------------|-----------|
| PM2.5     | 2.0         | 941.0      | Extreme pollution episodes |
| PM10      | 2.0         | 999.0      | Very high dust/smoke events |
| SO₂       | 0.2856      | 310.0      | Occasional industrial spikes |
| NO₂       | 1.85        | 270.0      | Heavy traffic peaks |
| CO        | 100         | 10,000     | Severe combustion events |
| O₃        | 0.21        | 1071.0     | Very strong photochemical smog on some days |

**Interpretation**: While average levels are already concerning, the **maximum values** show occasional **extreme pollution episodes**, typical of heavy winter haze in Beijing.

---

⚖️ **Quartiles (25%, 50%, 75%)**

**PM2.5**:
- 25% = 20.0 → Air was relatively clean 25% of the time.
- 50% (Median) = 55.0 → Half the time PM2.5 exceeded moderate levels.
- 75% = 111.0 → 25% of the time air quality was clearly unhealthy.

**PM10**:
- 25% = 35.0 | 50% = 81.0 | 75% = 142.0

**Interpretation**: The data is **right-skewed**. Most days show moderate-to-poor air quality, with frequent unhealthy episodes pushing the upper quartile high.

---

📊 **Standard Deviation (Variability)**

- PM2.5 std = **80.77**
- PM10 std = **91.40**
- CO std = **1167.62**
- O₃ std = **55.64**

**High standard deviation** across major pollutants indicates **strong variability** — pollution levels change dramatically between seasons (especially winter vs summer) and due to weather conditions.

---

🌦️ **Meteorological Summary**

- **Temperature**: Ranges from -16.8°C to 41.4°C (mean 13.57°C)
- **Pressure**: 982.4 – 1042.8 hPa
- **Wind Speed (WSPM)**: Mean 1.81 m/s (generally light winds, common in Beijing basin)
- **Rainfall**: Very low average (0.063 mm), confirming dry conditions most of the year

---

🟩 **Overall Interpretation**

This Beijing multi-station dataset reveals **persistently poor air quality** dominated by particulate matter (PM2.5 and PM10). Average PM2.5 levels (~80 µg/m³) are more than double China’s daily standard for "Good" air.

The data shows clear **seasonal patterns** with extreme spikes during winter months, likely due to heating, stagnant weather, and regional transport. While gaseous pollutants (SO₂, NO₂, CO) are moderate on average, they also show sharp occasional peaks.

The large range and high standard deviation highlight Beijing’s challenge with **severe episodic pollution events** alongside chronic background pollution.

This cleaned dataset is now well-prepared for time-series analysis, seasonal trend studies, and AQI categorization.